engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/definition/DefinitionExtractor.scala

```diff
@@ -67,13 +74,67 @@ object DefinitionExtractor {
 
   case class ObjectWithType(obj: Any, typ: TypingResult)
 
-  case class ObjectWithMethodDef(obj: Any,
+  trait ObjectWithMethodDef extends ObjectMetadata {
+
+    def invokeMethod(params: Map[String, Any],
+                     outputVariableNameOpt: Option[String],
+                     additional: Seq[AnyRef]) : Any
+
+    def objectDefinition: ObjectDefinition
+
+    def runtimeClass: Class[_]
+
+    def obj: Any
+
+    def annotations: List[Annotation]
+
+    override def parameters: List[Parameter] = objectDefinition.parameters
+
+    override def categories: List[String] = objectDefinition.categories
+
+    override def returnType: TypingResult = objectDefinition.returnType
+
+  }
+
+  abstract class OverriddenObjectWithMethodDef(original: ObjectWithMethodDef) extends ObjectWithMethodDef {
+
+    override def invokeMethod(params: Map[String, Any], outputVariableNameOpt: Option[String], additional: Seq[AnyRef]): Any
+
+    override def objectDefinition: ObjectDefinition = original.objectDefinition
+
+    override def runtimeClass: Class[_] = original.runtimeClass
+
+    override def obj: Any = original.obj
+
+    override def annotations: List[Annotation] = original.annotations
+  }
+
+  case class GenericNodeTransformationMethodDef(obj: GenericNodeTransformation[_], objectDefinition: ObjectDefinition) extends ObjectWithMethodDef {
+
+    override def invokeMethod(params: Map[String, Any], outputVariableNameOpt: Option[String], additional: Seq[AnyRef]): Any = {
+      val additionalParams = obj.nodeDependencies.map {
+        case TypedNodeDependency(klazz) =>
+          additional.find(klazz.isInstance).map(TypedNodeDependencyValue)
+            .getOrElse(throw new IllegalArgumentException(s"Failed to find dependency: $klazz"))
+        case OutputVariableNameDependency => outputVariableNameOpt.map(OutputVariableNameValue).getOrElse(throw new IllegalArgumentException("Output variable not defined"))
+        case _ => throw new IllegalArgumentException("Cannot handle this dependency...")
```

Comment:
Can you add "this" `other` in the message?

Reply:
:+1: added